google/cloud/bigtable/batcher.py

```diff
@@ -125,22 +298,120 @@ def mutate_rows(self, rows):
+        """Sends the current batch to Cloud Bigtable synchronously.
         For example:
 
         .. literalinclude:: snippets.py
             :start-after: [START bigtable_api_batcher_flush]
             :end-before: [END bigtable_api_batcher_flush]
             :dedent: 4
 
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+        rows_to_flush = []
+        while not self._rows.empty():
+            rows_to_flush.append(self._rows.get())
+        response = self.flush_rows(rows_to_flush)
+        return response
+
+    def flush_async(self):
+        """Sends the current batch to Cloud Bigtable asynchronously.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+
+        rows_to_flush = []
+        mutations_count = 0
+        mutations_size = 0
+        rows_count = 0
+        batch_info = BatchInfo()
+
+        while not self._rows.empty():
+            row = self._rows.get()
+            mutations_count += len(row._get_mutations())
+            mutations_size += row.get_mutations_size()
+            rows_count += 1
+            rows_to_flush.append(row)
+            batch_info.mutations_count = mutations_count
+            batch_info.rows_count = rows_count
+            batch_info.mutations_size = mutations_size
+
+            if (
+                rows_count >= self.flush_count
+                or mutations_size >= self.max_row_bytes
+                or mutations_count >= self.flow_control.max_mutations
+                or mutations_size >= self.flow_control.max_row_bytes
+                or self._rows.empty()  # submit when it reached the end of the queue
+            ):
+                self.flow_control.control_flow(batch_info)
+                future = self._executor.submit(self.flush_rows, rows_to_flush)
+                self.futures_mapping[future] = batch_info
+                future.add_done_callback(self.batch_completed_callback)
+
+                # reset and start a new batch
+                rows_to_flush = []
+                mutations_size = 0
+                rows_count = 0
+                mutations_count = 0
+                batch_info = BatchInfo()
+
+    def batch_completed_callback(self, future):
+        """Callback for when the mutation has finished.
+
+        Raise exceptions if there's any.
+        Release the resources locked by the flow control and allow enqueued tasks to be run.
+        """
+
+        processed_rows = self.futures_mapping[future]
+        self.flow_control.release(processed_rows)
+        del self.futures_mapping[future]
+
+    def flush_rows(self, rows_to_flush=None):
+        """Mutate the specified rows.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+        responses = []
+        if len(rows_to_flush) > 0:
```

Comment:
this would raise a `TypeError` if `flush_rows` is called without any arguments. Is that the intention?

Reply:
It was a mistake that the `rows_to_flush` was optional. I've changed it to required.